File: site/js/mods/transients.js
/**
 * Replace the Digitone II's FM drum transient bank with your own samples.
 *
 * The browser half of `src/dnfw/mods/transients.py`. The measurements are that
 * module's; repeated here only as constants, because a value copied without its
 * evidence is a value nobody can check:
 *
 * | | | how |
 * |---|---|---|
 * | region | DDR `0x8045b000`-`0x804ac000` | bounded by float32 tables either side, not a threshold |
 * | format | 16-bit mono, little-endian | byte-plane and step-correlation match real transients |
 * | rate | 48 kHz | the period is exactly 100.0 ms at 48 k, an awkward 108.8 ms at 44.1 k |
 * | entry | 4,800 samples | envelope autocorrelation, with 2x and 3x harmonics also ranking |
 * | phase | 2,496 samples in | attack-at-start vs decay-at-end scores 14.27 against 1.46 for the next |
 *
 * **Confirmed on hardware, 2026-09-14** (`docs/flashing.md`): an image with
 * markers at five known slots was flashed to a Digitone II and every marker
 * played back from `TRAN`.
 *
 * **What stays conditional:** none of this is documented by Elektron, so
 * another OS release could move the bank. The address is resolved through the
 * boot stream every time this runs, and an image whose layout is not the one it
 * was measured against is refused rather than written to.
 *
 * This replaces **whole entries in place** and touches nothing else -- no
 * lengths change, no index is rewritten, and every byte outside the entries it
 * writes is left exactly as it was.
 */

import { readSpan, spans } from "../bootstream.js";
import { replacement } from "../firmware.js";

export const ID = "transients";
export const NAME = "FM drum transients";
export const SECTION = 7;

const REGION_START = 0x8045b000;
const REGION_END = 0x804ac000;
const PHASE_SAMPLES = 2496;
export const ENTRY_SAMPLES = 4800;
export const RATE = 48000;

const BYTES_PER_SAMPLE = 2;
export const ENTRY_BYTES = ENTRY_SAMPLES * BYTES_PER_SAMPLE;
const BANK_ADDR = REGION_START + PHASE_SAMPLES * BYTES_PER_SAMPLE;
// `Math.floor`, not bare division: the region is 326,784 bytes, which is 34.04
// entries, and Python's `//` floors it. Without this COUNT is 34.04, which
// reads as 34 everywhere it is used as a length and as **35** in the bounds
// check below -- letting a write land past the end of the bank, in the float32
// table that bounds it. Caught by `scripts/js_mod_check.mjs` asserting the
// count rather than assuming the port was faithful.
export const COUNT = Math.floor((REGION_END - BANK_ADDR) / ENTRY_BYTES);   // 34

/**
 * The bank's file pieces, in load order. See `bootstream.spans`.
 *
 * **Not one range.** The bank is contiguous in the DSP's memory and scattered
 * in the file: two payload blocks with a 36-byte fill block and two 16-byte
 * headers between them, 47,440 bytes in. Reading or writing it linearly
 * crosses those headers, which is what this module did until 2026-09-14.
 */
function bankSpans(content) {
  return spans(content, BANK_ADDR, COUNT * ENTRY_BYTES);
}

/** The factory entries, each `ENTRY_BYTES` of raw 16-bit LE PCM. */
export function extract(firmware) {
  const section = firmware.container.find(SECTION);
  if (section === null) throw new Error(`image has no section ${SECTION}`);
  const content = section.unpack() ?? section.rawPayload;
  const bank = readSpan(content, BANK_ADDR, COUNT * ENTRY_BYTES);
  return Array.from({ length: COUNT }, (_, k) =>
    bank.subarray(k * ENTRY_BYTES, (k + 1) * ENTRY_BYTES));
}

/** One entry's samples as Float32 in [-1, 1), for preview and drawing. */
export function toFloat(entry) {
  const view = new DataView(entry.buffer, entry.byteOffset, entry.byteLength);
  const out = new Float32Array(entry.byteLength / 2);
  for (let i = 0; i < out.length; i++) out[i] = view.getInt16(i * 2, true) / 32768;
  return out;
}

/**
 * Float samples -> one slot's worth of 16-bit LE PCM, padded or truncated.
 *
 * Scaled by **32768 and clamped**, not by 32767. The two differ by one bit on
 * quiet material and nobody would hear it, but 32767 makes this the inexact
 * inverse of `toFloat`: a factory entry read out and written straight back
 * comes home altered, and the round-trip that is this mod's strongest check
 * would be testing the conversion instead of the bank. Clamping is what makes
 * 32768 safe -- it is the only value that would overflow.
 */
export function toEntry(samples) {
  const out = new Uint8Array(ENTRY_BYTES);
  const view = new DataView(out.buffer);
  const n = Math.min(samples.length, ENTRY_SAMPLES);
  for (let i = 0; i < n; i++) {
    const scaled = Math.round(samples[i] * 32768);
    view.setInt16(i * 2, Math.max(-32768, Math.min(32767, scaled)), true);
  }
  return out;
}

/**
 * Which bytes this mod writes, so a set of mods can be checked for overlap
 * before any of them runs (`src/dnfw/mods/__init__.py`).
 *
 * Byte-range overlap is the whole test, and it is the weakest useful
 * guarantee: two mods writing disjoint bytes can still fight over the same
 * feature and nothing here detects that.
 */
export function extents(firmware) {
  const section = firmware.container.find(SECTION);
  const content = section.unpack() ?? section.rawPayload;
  return bankSpans(content)
    .filter((piece) => piece.at !== null)
    .map((piece) => ({
      section: SECTION,
      start: piece.at,
      length: piece.length,
      what: `transient bank, ${piece.length.toLocaleString()} bytes`,
    }));
}

/**
 * Replace entries with `entries` -- a sparse Map of slot index -> Uint8Array of
 * `ENTRY_BYTES`. Slots absent from the map keep their factory bytes.
 *
 * Sparse on purpose: replacing one transient out of 34 is the common case, and
 * an API taking a full list would make a user supply 33 samples they do not
 * care about, or make the tool invent them.
 *
 * -> a Section ready for `firmware.build`.
 */
export function apply(firmware, entries) {
  const section = firmware.container.find(SECTION);
  if (section === null) throw new Error(`image has no section ${SECTION}`);
  const original = section.unpack() ?? section.rawPayload;
  const content = original.slice();       // a copy: the loaded image stays clean

  // Build the whole bank in load order, then scatter it back across the blocks
  // that hold it. Writing entry by entry at a computed file offset is what put
  // header bytes in the audio, and would have written audio over two block
  // headers -- leaving a boot stream the DSP cannot load.
  const bank = readSpan(content, BANK_ADDR, COUNT * ENTRY_BYTES);

  const notes = [];
  for (const [slot, entry] of entries) {
    if (!Number.isInteger(slot) || slot < 0 || slot >= COUNT) {
      throw new Error(`slot ${slot} is outside 0..${COUNT - 1}`);
    }
    if (entry.length !== ENTRY_BYTES) {
      throw new Error(
        `slot ${slot}: ${entry.length} bytes, expected exactly ${ENTRY_BYTES}`);
    }
    bank.set(entry, slot * ENTRY_BYTES);
    notes.push(`${String(slot).padStart(2, "0")} replaced`);
  }

  let cursor = 0;
  for (const { at, length } of bankSpans(content)) {
    if (at === null) {
      // A fill block has no file bytes, so whatever the user put here cannot
      // be stored. Reported rather than dropped in silence: it is 36 bytes
      // near the end of entry 4, and the loader zeroes them regardless.
      if (bank.subarray(cursor, cursor + length).some((b) => b !== 0)) {
        notes.push(
          `${length} bytes at bank offset ${cursor.toLocaleString()} (entry `
          + `${Math.floor(cursor / ENTRY_BYTES)}) fall in a boot-stream fill `
          + "block and CANNOT be written; the loader zeroes them");
      }
    } else {
      content.set(bank.subarray(cursor, cursor + length), at);
    }
    cursor += length;
  }

  return {
    section: replacement(firmware, SECTION, content),
    notes,
    extents: extents(firmware),
  };
}

File: src/dnfw/mods/transients.py
"""Replace the Digitone II's FM drum transient bank with your own samples.

## What was measured, and what was not

The bank was located in the **SHARC image** (container section 7, an ADI boot
stream) by a detector calibrated against Elektron's own `TRANSIENT 01-08.wav`
from the Syntakt OS 1.41 Twinshot pack — `docs/pcm-hunt.md` §14. The numbers
below are measurements, not choices:

| | | how |
|---|---|---|
| region | DDR `0x8045b000`–`0x804ac000` | bounded by float32 tables either side, not by a threshold |
| format | 16-bit mono, little-endian | byte-plane and step-correlation match real transients |
| rate | 48 kHz | the period is exactly 100.0 ms at 48 k and an awkward 108.8 ms at 44.1 k |
| entry | 4,800 samples | envelope autocorrelation, with 2x and 3x harmonics also ranking |
| phase | 2,496 samples into the region | attack-at-start vs decay-at-end scores 14.27 against 1.46 for the next candidate |

**Confirmed on hardware, 2026-09-14.** An image with marker samples at five
known slots was flashed to a Digitone II and every marker played back from
`TRAN` — so these bytes are the FM drum transients, and the count is 34.
`docs/pcm-hunt.md` §16.

**What stays conditional:** none of this is documented by Elektron, so another
OS release could move the bank. `_payload_offset` resolves the address through
the boot stream every time it runs and refuses an image whose layout it does not
recognise, rather than trusting the constants above.

This mod replaces **whole entries in place** and touches nothing else — no
lengths change, no index is rewritten, and every byte outside the entries it
writes is left exactly as it was.

## How a replacement is prepared

Each input becomes exactly 4,800 samples of 16-bit mono at 48 kHz:

- other sample rates are linearly resampled;
- stereo is mixed to mono;
- longer inputs are **truncated** and shorter ones **zero-padded**, because a
  transient is a fixed slot in a packed bank and there is nowhere for a longer
  one to go.

Truncation is reported rather than done silently.
"""

import pathlib
import struct
import wave

from . import Extent, ModError, Result
from ..image import bootstream

ID = "transients"
NAME = "FM drum transients"
SUMMARY = "Replace the factory FM drum transient bank with your own samples."
DEVICE = 0x15                      # Digitone II, dump-protocol product id
SECTION = 7                        # the SHARC image

# Measured -- see the module docstring. Load addresses, not file offsets.
REGION_START = 0x8045B000
REGION_END = 0x804AC000
PHASE_SAMPLES = 2496
ENTRY_SAMPLES = 4800
RATE = 48000

BYTES_PER_SAMPLE = 2
ENTRY_BYTES = ENTRY_SAMPLES * BYTES_PER_SAMPLE
BANK_ADDR = REGION_START + PHASE_SAMPLES * BYTES_PER_SAMPLE
COUNT = (REGION_END - BANK_ADDR) // ENTRY_BYTES


def _payload_offset(section_bytes: bytes, address: int) -> int:
    """Load address -> offset inside section 7's unpacked payload.

    The boot stream scatters its payloads across L1, L2 and DDR, so an offset
    into the section is not an address (`dnfw.image.bootstream`). Writing to
    the wrong one would corrupt an unrelated region, so the block that actually
    covers the address is found rather than assumed.
    """
    walk = bootstream.walk(section_bytes)
    for block in walk.blocks:
        if not block.has_payload:
            continue
        if block.target <= address < block.target + block.count:
            return block.payload_at + (address - block.target)
    raise ModError(
        f"0x{address:08x} is not inside any boot-stream block carrying data; "
        f"this image's layout is not the one this mod was measured against")


def extents(firmware) -> list[Extent]:
    section = firmware.container.find(SECTION)
    if section is None:
        raise ModError(f"image has no section {SECTION}")
    data = section.unpack() or section.raw_payload
    start = _payload_offset(data, BANK_ADDR)
    return [Extent(SECTION, start, COUNT * ENTRY_BYTES,
                   f"{COUNT} transient entries of {ENTRY_SAMPLES} samples")]


def read_wav(path: pathlib.Path) -> list[int]:
    """-> exactly ENTRY_SAMPLES signed 16-bit samples at RATE, mono."""
    with wave.open(str(path), "rb") as w:
        if w.getsampwidth() != 2:
            raise ModError(f"{path.name}: {w.getsampwidth() * 8}-bit; "
                           f"only 16-bit WAV is supported")
        ch, rate, n = w.getnchannels(), w.getframerate(), w.getnframes()
        raw = w.readframes(n)
    v = list(struct.unpack_from(f"<{n * ch}h", raw, 0))
    if ch > 1:
        v = [sum(v[i:i + ch]) // ch for i in range(0, len(v), ch)]
    if rate != RATE:
        # Linear resample. Good enough for a 100 ms percussive hit, and far
        # better than refusing the file; anything fancier belongs in the
        # user's editor, where they can hear it.
        out, step = [], rate / RATE
        for i in range(int(len(v) / step)):
            x = i * step
            k = int(x)
            f = x - k
            a = v[k] if k < len(v) else 0
            b = v[k + 1] if k + 1 < len(v) else a
            out.append(int(a + (b - a) * f))
        v = out
    if len(v) > ENTRY_SAMPLES:
        v = v[:ENTRY_SAMPLES]
    elif len(v) < ENTRY_SAMPLES:
        v = v + [0] * (ENTRY_SAMPLES - len(v))
    return [max(-32768, min(32767, x)) for x in v]


def prepare(path: pathlib.Path, lead_ms: int = 3) -> list[int]:
    """Condition any sample into one slot: mono, 48 kHz, onset-aligned, 100 ms.

    A transient slot is 100 ms and a user's file is whatever it is, so something
    has to choose *which* 100 ms. Starting at sample 0 is wrong for most
    recordings: files routinely carry a few milliseconds of silence, and that
    silence would eat the attack, which is the one part of a transient that
    matters.

    So the onset is found -- the first point reaching a fraction of the peak --
    and the slot starts `lead_ms` before it, keeping the attack's leading edge
    rather than clipping into it. A short fade at the end prevents the hard cut
    from adding a click of its own, which would be a transient this tool
    invented.

    This is deliberately NOT transient/tonal separation. Pulling the percussive
    layer out of a pitched sample is a real DSP problem, and
    `mikkovihonen/transientsplit` (MIT) already solves it in the browser -- run
    a sample through that first if you want the transient *component*, then
    through this to fit it to the slot.
    """
    v = read_wav(path)                      # mono, 48 kHz, already slot-length
    peak = max((abs(x) for x in v), default=0)
    if peak == 0:
        return v
    threshold = peak // 8
    onset = next((i for i, x in enumerate(v) if abs(x) >= threshold), 0)
    lead = lead_ms * RATE // 1000
    start = max(0, onset - lead)
    out = v[start:] + [0] * start
    fade = RATE // 500                      # 2 ms, far shorter than any decay
    for i in range(min(fade, len(out))):
        out[-1 - i] = int(out[-1 - i] * i / fade)
    return out


def extract(firmware) -> list[bytes]:
    """-> the factory entries, each ENTRY_BYTES of raw 16-bit LE PCM."""
    section = firmware.container.find(SECTION)
    data = section.unpack() or section.raw_payload
    start = _payload_offset(data, BANK_ADDR)
    return [data[start + k * ENTRY_BYTES: start + (k + 1) * ENTRY_BYTES]
            for k in range(COUNT)]


def apply(firmware, sources: list[pathlib.Path], condition: bool = False) -> Result:
    """Replace entries in order with `sources`; a short list leaves the rest.

    `condition=True` runs each input through `prepare()` -- onset alignment and
    an end fade. It is OFF by default on purpose: conditioning changes samples,
    and with it off, extracting the factory bank and writing it straight back
    produces a byte-identical image. That round-trip is the strongest check this
    mod has, and a default that silently altered samples would destroy it.
    """
    if not sources:
        raise ModError("no input samples given")
    if len(sources) > COUNT:
        raise ModError(f"{len(sources)} samples given but the bank holds "
                       f"{COUNT}; remove {len(sources) - COUNT}")

    section = firmware.container.find(SECTION)
    if section is None:
        raise ModError(f"image has no section {SECTION}")
    data = bytearray(section.unpack() or section.raw_payload)
    start = _payload_offset(bytes(data), BANK_ADDR)

    notes = []
    for k, path in enumerate(sources):
        with wave.open(str(path), "rb") as w:
            n_in, rate_in = w.getnframes(), w.getframerate()
        samples = prepare(path) if condition else read_wav(path)
        at = start + k * ENTRY_BYTES
        data[at:at + ENTRY_BYTES] = struct.pack(f"<{ENTRY_SAMPLES}h", *samples)
        note = f"{k:02d} <- {path.name}"
        if condition:
            note += " (onset-aligned)"
        if rate_in != RATE:
            note += f" (resampled {rate_in} -> {RATE})"
        if n_in * RATE // max(rate_in, 1) > ENTRY_SAMPLES:
            note += f" (TRUNCATED to {1000 * ENTRY_SAMPLES // RATE} ms)"
        notes.append(note)
    if len(sources) < COUNT:
        notes.append(f"entries {len(sources)}..{COUNT - 1} left as factory")

    return Result(payloads={SECTION: bytes(data)},
                  extents=[Extent(SECTION, start, COUNT * ENTRY_BYTES,
                                  "transient bank")],
                  notes=notes)
